pypdfium2 4.19.0

boldorider4 opened the pull request:
Jira: [PKG-2559](https://anaconda.atlassian.net/browse/PKG-2559)
Upstream project: pypdfium2 [requirements](https://github.com/pypdfium2-team/pypdfium2/blob/4.18.0/pyproject.toml)

[PKG-2559]: https://anaconda.atlassian.net/browse/PKG-2559?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

boldorider4 committed:
add skeleton recipe

boldorider4 committed:
complete recipe

boldorider4 committed:
do not specify lower bound for python

boldorider4 committed:
include every license file

boldorider4 committed:
set url to proper tar gz address

boldorider4 committed:
remove noarch

boldorider4 committed:
pin ctypesgen to pypdfium2-team specific version

boldorider4 committed:
fix pinning

anaconda-pkg-build commented:
<details><summary>Linter check found the following problems:</summary>
The following problems have been found:

ERROR: clone/recipe/meta.yaml:22: host_section_needs_exact_pinnings: Packages in host must have exact version pinnings, except python build tools.
INFO: clone/recipe/meta.yaml:50: invalid_url: https://pypdfium2.readthedocs.io/en/stable : URL domain redirect pypdfium2.readthedocs.io ->
ERROR: clone/recipe/meta.yaml:51: missing_license_family: The recipe is missing the `about/license_family` key.
WARNING: clone/recipe/meta.yaml:51: missing_license_file: The recipe is missing the `about/license_file` or `about/license_url` key.
ERROR: clone/recipe/meta.yaml:51: missing_home: The recipe is missing a homepage URL
Errors were found

</details>

boldorider4 committed:
fix version pinning

anaconda-pkg-build commented:
<details><summary>Linter check found the following problems:</summary>
The following problems have been found:

ERROR: clone/recipe/meta.yaml:22: host_section_needs_exact_pinnings: Packages in host must have exact version pinnings, except python build tools.
INFO: clone/recipe/meta.yaml:50: invalid_url: https://pypdfium2.readthedocs.io/en/stable : URL domain redirect pypdfium2.readthedocs.io ->
WARNING: clone/recipe/meta.yaml:51: missing_license_file: The recipe is missing the `about/license_file` or `about/license_url` key.
ERROR: clone/recipe/meta.yaml:51: missing_home: The recipe is missing a homepage URL
ERROR: clone/recipe/meta.yaml:51: missing_license_family: The recipe is missing the `about/license_family` key.
Errors were found

</details>

boldorider4 committed:
add git and c compiler to build

anaconda-pkg-build commented:
<details><summary>Linter check found the following problems:</summary>
The following problems have been found:

ERROR: clone/recipe/meta.yaml:25: host_section_needs_exact_pinnings: Packages in host must have exact version pinnings, except python build tools.
INFO: clone/recipe/meta.yaml:53: invalid_url: https://pypdfium2.readthedocs.io/en/stable : URL domain redirect pypdfium2.readthedocs.io ->
ERROR: clone/recipe/meta.yaml:54: missing_license_family: The recipe is missing the `about/license_family` key.
WARNING: clone/recipe/meta.yaml:54: missing_license_file: The recipe is missing the `about/license_file` or `about/license_url` key.
ERROR: clone/recipe/meta.yaml:54: missing_home: The recipe is missing a homepage URL
Errors were found

</details>

boldorider4 committed:
pin pdfium-binaries to host and run

anaconda-pkg-build commented:
<details><summary>Linter check found the following problems:</summary>
The following problems have been found:

ERROR: clone/recipe/meta.yaml:26: host_section_needs_exact_pinnings: Packages in host must have exact version pinnings, except python build tools.
INFO: clone/recipe/meta.yaml:55: invalid_url: https://pypdfium2.readthedocs.io/en/stable : URL domain redirect pypdfium2.readthedocs.io ->
WARNING: clone/recipe/meta.yaml:56: missing_license_file: The recipe is missing the `about/license_file` or `about/license_url` key.
ERROR: clone/recipe/meta.yaml:56: missing_home: The recipe is missing a homepage URL
ERROR: clone/recipe/meta.yaml:56: missing_license_family: The recipe is missing the `about/license_family` key.
Errors were found

</details>

boldorider4 committed:
fix build script

boldorider4 committed:
patch upstream project so it can be built as a conda package

boldorider4 committed:
add pillow as run dependency

anaconda-pkg-build commented:
<details><summary>Linter check found the following problems:</summary>
The following problems have been found:

ERROR: clone/recipe/meta.yaml:31: host_section_needs_exact_pinnings: Packages in host must have exact version pinnings, except python build tools.
ERROR: clone/recipe/meta.yaml:62: missing_license_family: The recipe is missing the `about/license_family` key.
WARNING: clone/recipe/meta.yaml:62: missing_license_file: The recipe is missing the `about/license_file` or `about/license_url` key.
ERROR: clone/recipe/meta.yaml:62: missing_home: The recipe is missing a homepage URL
Errors were found

</details>

mara004 commented:
I just viewed the patch. Please see https://github.com/bblanchon/pdfium-binaries/issues/119#issuecomment-1697474571 for an idea how your goal could be achieved already without any changes to pypdfium2's code.

Probably you have realized this already, but FWIW the current patch (f869bfa0c381ad01df348e6792dd7e0b20bf6301) is thoroughly inappropriate. You really should not go commenting out large parts of intentional code. This completely breaks the intended use case, so it is not mergeable. And it is also not maintainable in this feedstock. Instead, if you want to run a different code passage, this can simply be controlled through the top level `setup.py`, as is implemented already with `$PDFIUM_PLATFORM=none`. Sorry for writing so bluntly. As the person who wrote the code, it's just a bit of a shock to see such a patch. 😅

PS: The ctypesgen call style improvement of the patch is good, but was already implemented independently in `devel_v5`.

mara004 commented:
BTW, regarding commit 7dbe15650ea73a56667ca2cd38948a1fb5585b45, pillow is meant to be optional at runtime, only a subset of features depends on it (e.g. the rendering CLI). So no need to specify it as hard dependency.
A related oversight introduced in the devel_v5 branch was recently fixed, see https://github.com/pypdfium2-team/pypdfium2/pull/249. `main` should be OK though.

boldorider4 committed:
add abs.yaml

anaconda-pkg-build commented:
<details><summary>Linter check found the following problems:</summary>
The following problems have been found:

ERROR: clone/recipe/meta.yaml:31: host_section_needs_exact_pinnings: Packages in host must have exact version pinnings, except python build tools.
WARNING: clone/recipe/meta.yaml:62: missing_license_file: The recipe is missing the `about/license_file` or `about/license_url` key.
ERROR: clone/recipe/meta.yaml:62: missing_home: The recipe is missing a homepage URL
ERROR: clone/recipe/meta.yaml:62: missing_license_family: The recipe is missing the `about/license_family` key.
Errors were found

</details>

boldorider4 committed:
package reference binding on linux

anaconda-pkg-build commented:
<details><summary>Linter check found the following problems:</summary>
The following problems have been found:

ERROR: clone/recipe/meta.yaml:32: host_section_needs_exact_pinnings: Packages in host must have exact version pinnings, except python build tools.
ERROR: clone/recipe/meta.yaml:63: missing_license_family: The recipe is missing the `about/license_family` key.
WARNING: clone/recipe/meta.yaml:63: missing_license_file: The recipe is missing the `about/license_file` or `about/license_url` key.
ERROR: clone/recipe/meta.yaml:63: missing_home: The recipe is missing a homepage URL
Errors were found

</details>

boldorider4 committed:
remove abs.yaml

boldorider4 committed:
skip archs for which no pdfium binaries exist

anaconda-pkg-build commented:
<details><summary>Linter check found the following problems:</summary>
The following problems have been found:

ERROR: clone/recipe/meta.yaml:33: host_section_needs_exact_pinnings: Packages in host must have exact version pinnings, except python build tools.
WARNING: clone/recipe/meta.yaml:64: missing_license_file: The recipe is missing the `about/license_file` or `about/license_url` key.
ERROR: clone/recipe/meta.yaml:64: missing_license_family: The recipe is missing the `about/license_family` key.
ERROR: clone/recipe/meta.yaml:64: missing_home: The recipe is missing a homepage URL
Errors were found

</details>

boldorider4 committed:
address linter

M-Waszkiewicz-Anaconda committed:
fix windows build

M-Waszkiewicz-Anaconda committed:
regenerate patches

boldorider4 commented on recipe/0003-fix-windows-include-location.patch:
```diff
@@ -1,37 +1,37 @@
+         # now passing $SRC_DIR/data/<arch> and $PREFIX
+-        call_ctypesgen(pl_dir, prefix_dir/"include")
++        include_dir = prefix_dir/"include"
 +        if platform.system().lower() == "windows":
```
You can also remove the if condition and use the `# [win]` selector in the recipe for this patch only.

M-Waszkiewicz-Anaconda replied:
yes, I could but I think this is better, by applying patch for different platform you constructing 2 different branches of source code

anaconda-pkg-build commented:
<details><summary>Linter check found the following problems:</summary>
The following problems have been found:

ERROR: clone/recipe/meta.yaml:26: no_git_on_windows: git should not be used as a dependency on Windows.
Errors were found

</details>

M-Waszkiewicz-Anaconda committed:
use latest ctypesgen-pypdfium2-team

M-Waszkiewicz-Anaconda committed:
Revert "use latest ctypesgen-pypdfium2-team"

This reverts commit 6199efbacd3e02baba9e1393f598fd26c94e603d.

M-Waszkiewicz-Anaconda committed:
use latest ctypesgen-pypdfium2-team

M-Waszkiewicz-Anaconda committed:
add abs.yaml

JeanChristopheMorinPerso commented on recipe/meta.yaml:
```diff
@@ -0,0 +1,73 @@
+    - python
+    - setuptools
+    - pip
+    - build
```
I have a feeling this isn't required.

mara004 replied:
I think you're right. `build` isn't required here. This error was inherited from upstream, so my fault basically. 😅 (IIRC `build` is only called as subprocess in `craft_packages.py`, which runs in our workflow, but not on user setup. So it doesn't belong in the pyproject build-system requires, nor here.)

M-Waszkiewicz-Anaconda replied:
I will remove the build

JeanChristopheMorinPerso commented on recipe/meta.yaml:
```diff
@@ -0,0 +1,73 @@
+requirements:
+  build:
+    - git         # [not win]
+    - {{ compiler('c') }}
```
In the build logs, I see
```
INFO: Status: Preprocessing /tmp/tmprlu8i60_.h
INFO: Status: gcc -E -U __GNUC__ -dD "-D__extension__=" "-D__const=const" "-D__asm__(x)=" "-D__asm(x)=" "-DCTYPESGEN=1" "/tmp/tmprlu8i60_.h"
ERROR: gcc -E: /bin/sh: gcc: command not found
```

mara004 replied:
This might currently be masked by the patch to use the reference bindings?
I didn't check if the reference bindings code is truly active, but if it is, then the ctypesgen call's result would be unused. If it is not and no C preprocessor was found, the import test should fail.

M-Waszkiewicz-Anaconda replied:
what we should do exactly ? Import test is passing, so the build is doing something which is not needed,  but that would not affect end user

mara004 replied:
> Import test is passing, so the build is doing something which is not needed, but that would not affect end user

Yes, this would be my understanding.

> what we should do exactly ?

Certainly not what you are currently doing. The whole patching is thoroughly messed up, unsustainable and should be revised from the ground, as stated before: https://github.com/AnacondaRecipes/pypdfium2-feedstock/pull/2#issuecomment-1698031058, https://github.com/bblanchon/pdfium-binaries/issues/119#issuecomment-1697474571

I would recommend to just use `$PDFIUM_PLATFORM=none` and deploy the bindings externally in the recipe's script section so you don't need any patches in the first place.

However, if you really do want to run custom code in our setup infra, you should start at the top-level `setup.py` by adding a new case in `packaging_handler()`, e.g. `elif target == "conda-thirdparty"` and do your work there (deploying bindings & binary into `src/pypdfium2/` and calling `setuptools.setup(...)` at the end).
Then you would define `PDFIUM_PLATFORM="conda-thirdparty"` in the recipe env to use this new code passage.
If done correctly, the patch should effectively be additions-only and really short.

Apart from that, I'm wondering where you currently get the binary from? I suppose it's ctypesgen's library loader finding it automatically from the env's sysroot, where supposedly the pdfium-binaries are installed to?
This should ideally be made explicit by symlinking or passing ctypesgen the right runtime libdirs.

M-Waszkiewicz-Anaconda replied:
I have just jumped in to make windows build working, I do not have a strong opinion on what @boldorider4  did, but from what I see it is working and all tests passed and this package is not going into public but rather private channel, so maybe it is good enough to go ahead with it

mara004 replied:
Well, the patch will be a burden to maintain, for one thing.
If this is a "rebuild once in a blue moon" project, then maybe it's doable, but technically it's just a terrible mess, while it should be comparatively easy to get right.

mara004 replied:
Ah, I just noticed @skupr-anaconda changed it to use the approach I suggested above. That looks substantially better, thanks!

M-Waszkiewicz-Anaconda committed:
remove the build package requirement

skupr-anaconda committed:
Update to 4.19.0:

- Add PDFIUM_PLATFORM=none

- Add  to script

- Disable patches

- Disable git in build

- Disable pillow in run

M-Waszkiewicz-Anaconda committed:
add pillow, clean up

M-Waszkiewicz-Anaconda reviewed (approved):
LGTM

skupr-anaconda reviewed (approved):
LGTM

mara004 commented on recipe/meta.yaml:
```diff
@@ -0,0 +1,69 @@
+  script_env:
+    - PDFIUM_PLATFORM=none
+  script:
+    - cp bindings/raw.py src/pypdfium2/raw.py
```
This is indeed much simpler. Thanks!

mara004 commented on recipe/meta.yaml:
```diff
@@ -0,0 +1,69 @@
+    - setuptools
+    - pip
+    - wheel !=0.38.0,!=0.38.1
+    - ctypesgen-pypdfium2-team 1.1.1+17_g10b2b4
```
If using the reference bindings, then you don't need ctypesgen here anymore.
(The wheel code passage is also not used with `PDFIUM_PLATFORM=none`, but it is currently still an import-time requirement. We could change this, though.)

mara004 commented on recipe/meta.yaml:
```diff
@@ -0,0 +1,69 @@
+    - python
+    - pdfium-binaries 118.0.5868
+    # to run pypdfium2 imgtopdf you need to have pillow
+    - pillow
```
As mentioned in https://github.com/AnacondaRecipes/pypdfium2-feedstock/pull/2#issuecomment-1698164946, pillow is supposed to be optional, only a subset of features depends on it.
i.e. it's not strictly necessary for all use cases. But it may still be OK to keep since it's very commonly needed (upstream, we got repeated questions why pillow is not in the pyproject dependencies).

mara004 commented on recipe/meta.yaml:
```diff
@@ -0,0 +1,69 @@
+    - pdfium-binaries 118.0.5868
+  run:
+    - python
+    - pdfium-binaries 118.0.5868
```
mismatched for v4.19, you'll want 5975 instead

~~Apart from that, it's a problem the hardcoded pdfium version of our release remains, we should set that to unknown when using `PDFIUM_PLATFORM=none`. Will fix this upstream soon.~~ fixed

M-Waszkiewicz-Anaconda committed:
update pdfium-binaries dependency

JeanChristopheMorinPerso commented on recipe/meta.yaml:
```diff
@@ -0,0 +1,68 @@
+
+requirements:
+  build:
+    - {{ compiler('c') }}
```
Is this still needed?

mara004 replied:
no.

JeanChristopheMorinPerso commented on recipe/meta.yaml:
```diff
@@ -0,0 +1,68 @@
+  license_file:
+    - LICENSES/Apache-2.0.txt
+    - LICENSES/BSD-3-Clause.txt
+    - LICENSES/LicenseRef-PdfiumThirdParty.txt
```
Isn't this required only when vendoring pdfium?

mara004 replied:
Yes, `LicenseRef-PdfiumThirdParty.txt` is not needed here.

Something similar is required in the pdfium-binaries package, though, but I think you already have this by bundling the license file from upstream tarballs.

M-Waszkiewicz-Anaconda committed:
remove compiler, pdfium license

mara004 commented:
The current source URL `https://pypi.io/packages/source/{{ name[0] }}/{{
name }}/{{ name }}-{{ version }}.tar.gz` seems strange.
Is it really dynamic `{{ name[0] }}` or shouldn't it be just `p`?
That said, I think it would be better to use the GH release URL, as you had
initially.

On Tue, Sep 5, 2023 at 5:19 PM Serhii Kupriienko ***@***.***>
wrote:

> ***@***.**** approved this pull request.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/AnacondaRecipes/pypdfium2-feedstock/pull/2#pullrequestreview-1611311278>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/APW4VW6VMTJ5PSULVHS6YPLXY47JFANCNFSM6AAAAAA3UVHXX4>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>

JeanChristopheMorinPerso commented:
There is a GH outage (https://www.githubstatus.com/incidents/smdz34v7j8q0) so I can't hit the approve button. But I approve.
